src/write_gcm.c

```diff
@@ -3823,7 +3828,8 @@ static int wg_transmit_unique_segment(const wg_context_t *ctx,
 
     // By the way, a successful response is an empty JSON record (i.e. "{}").
     // An unsuccessful response is a detailed error message from the API.
-    char response[2048];
+    write_ctx.size = 0;
```

Comment:
I don't agree. The whole point is we're *reusing* the allocation from the previous iteration. We only ```sfree``` on the way out of the function

Reply:
Could be worth a comment here though.

Reply:
It's fine to reuse the allocation, but not the contents. You're right that `sfree` is overkill, but we at least need `write_ctx.data[0] = '\0'` (given that we'll switch to an empty string as the default).

Reply:
Is this C trickery where I can set the 0th element of a NULL array to `'\0'`? The first iteration, .data is set to NULL.

Reply:
Correct. You change `write_ctx.data` to an empty string (null terminated) without actually having to empty the string or lose the pointer to the buffer.

Reply:
Which choice are you suggesting?
`write_ctx.data[0] = '\0';` or `write_ctx.data = '\0';`?

Reply:
The first one crashes due to this error:
```==25952== Invalid write of size 1
==25952==    at 0xA0B24B4: wg_transmit_unique_segment (write_gcm.c:3831)
==25952==    by 0xA0B3C09: wg_transmit_unique_segments (write_gcm.c:3771)
==25952==    by 0xA0B3C09: wg_process_queue (write_gcm.c:3594)
==25952==    by 0x4E3F183: start_thread (pthread_create.c:312)
==25952==    by 0x535CFFC: clone (clone.S:111)
==25952==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
```

The second one "seems" like it would leak, due to reassigning the value entirely without freeing the previous contents.

Reply:
No, there is no such C trickery. Your intuition that a null pointer dereference would cause a segfault is correct. (Though that's a nice error message... where did you get that? gdb?)

Anyway, there are so many cooks in this conversation that I'm sure it must feel horribly frustrating. My personal take is that these are the objectives:
1. Don't leak memory
1. Don't use fixed-size buffers
1. Have a nice-looking API
1. Have the code look nice and readable
1. Reuse the buffer if it's convenient

I don't think I agree with @igorpeshansky's suggestion to set ```write_ctx.data[0]='\0'```

If---and note this is an IF---you want to go with the idea of getting ```write_ctx_t``` out of the API, then I think the contract for ```wg_curl_get_or_post``` is simply:
1. It takes the address of a ```char*``` pointer (i.e. a ```char**```, which acts as an in/out parameter)
2. On the way in, the *caller* promises that the ```char*``` pointer is either NULL or it points to a valid buffer (allocated by ```malloc``` or ```realloc```), but with no guarantee about contents and no assumptions about capacity and specifically no guarantee about whether it's a 0-terminated string or not.
3. On the way out, the *callee* promises that the buffer is either
* NULL, which indicates a malloc or realloc error.
* a pointer to allocated storage which contains a \0-terminated string (this could be an empty or non-empty string, i.e. maybe the very first byte is 0)

Under the above rules there is no need for the *caller* to "condition" the buffer by making sure it's 0-terminated on the way in, so there's no need for the ```write_ctx.data[0]='\0'```

If one were to accept the above, there are a couple of subpoints to argue about:
1. Upon realloc failure (which should never happen unless the server sends us multi-gigabyte repsonses), should the caller get back some prefix of the response or a NULL. If the caller wants some prefix of the string then the above rules change somewhat and there should be an indication passed to the caller that the response was truncated (so its JSON parser doesn't freak out or whatever). I personally don't see the value of handling a partial response, as all I would do is write to the log ("returned string too huge, sorry, truncated"). But if you live in a world where ANY malloc or realloc can fail, then the caller *still* has to be prepared to get a NULL back, a partial response back, or a complete response back. I think you can reduce those cases by not giving back the partial response. If a NULL comes back then some realloc failed and the caller doesn't know whether it was the first or the tenth.

Reply:
Corey, the error is from `valgrind`.
Bryan, the proposal to set `write_ctx.data[0]` to `'\0'` would only work if we initialize it to `malloc(1)`, the way you had it before. We were discussing doing this anyway, to distinguish from a failed allocation, which would return `NULL`.

To answer Corey's point, on `realloc` failure we want a prefix, not a `NULL` -- these are errors coming back from the server, and it's better to have some than none.